### PR TITLE
Add support for renaming column in Delta Lake

### LINKED ISSUE #13538: Add support for `RENAME COLUMN` in Delta Lake connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Fixes https://github.com/trinodb/trino/issues/12638
Fixes #13538

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Add support for renaming columns. ({issue}`15821 `)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
`web-ui-checks` hit #16179

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java [hunk 1]
```diff
@@ -331,6 +330,33 @@ public void testDropNonEmptySchemaWithTable()
         assertUpdate("DROP SCHEMA " + schemaName);
     }
 
+    @Override
+    public void testRenameColumn()
+    {
+        // Override because the connector doesn't support renaming columns with 'none' column mapping
+        // There are some tests in in io.trino.tests.product.deltalake.TestDeltaLakeColumnMappingMode
+        assertThatThrownBy(super::testRenameColumn)
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java:338]: It would be beneficial to have https://github.com/trinodb/trino/pull/17565 so that we can properly create such tables.
Alternatively, `register_table` would be also an option (not my prefferred option).

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java [hunk 2]
```diff
@@ -1188,6 +1188,158 @@ public Object[][] columnMappingWithTrueAndFalseDataProvider()
...
+
+            // Ensure renaming to the dropped column doesn't restore the old data
+            onDelta().executeQuery("ALTER TABLE default." + tableName + " DROP COLUMN id");
+            renameColumns.accept(tableName, new RenameColumn("new_data", "id"));
+
+            assertThat(onTrino().executeQuery("SELECT id, new_part FROM delta.default." + tableName))
+                    .containsOnly(row(10, "part#1"));
+            assertThat(onDelta().executeQuery("SELECT id, new_part FROM default." + tableName))
+                    .containsOnly(row(10, "part#1"));
+        }
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:1250]: Can we have a basic set of DML operations dealing with the renamed column to ensure their accuracy when dealing with a renamed column?
Actually, this should be achieved as an integration test (and not a product test) with a table registered from resources directory

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 3]
```diff
@@ -208,6 +209,58 @@ public void testAddNestedColumnWithColumnMappingMode(String columnMappingMode)
...
+        JsonNode schema = OBJECT_MAPPER.readTree(metadata.getSchemaString());
+        List<JsonNode> fields = ImmutableList.copyOf(schema.get("fields").elements());
+        Assertions.assertThat(fields).hasSize(2);
+        JsonNode integerColumn = fields.get(0);
+        JsonNode nestedColumn = fields.get(1);
+        List<JsonNode> rowFields = ImmutableList.copyOf(nestedColumn.get("type").get("fields").elements());
+        Assertions.assertThat(rowFields).hasSize(3);
+
+        // Rename 'second_col' column and verify that nested metadata are same except for 'name' field and the table configuration are preserved
+        assertUpdate("ALTER TABLE " + tableName + " RENAME COLUMN second_col TO renamed_col");
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:246]: Can we ensure that the `COMMENT` of the column is also there?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 4]
```diff
@@ -1332,6 +1334,84 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
 
+    @Override
+    public void renameColumn(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle, String newColumnName)
+    {
+        DeltaLakeTableHandle table = (DeltaLakeTableHandle) tableHandle;
+        DeltaLakeColumnHandle deltaLakeColumn = (DeltaLakeColumnHandle) columnHandle;
+        String sourceColumnName = deltaLakeColumn.getBaseColumnName();
+
+        checkSupportedWriterVersion(session, table);
+        if (changeDataFeedEnabled(table.getMetadataEntry()) && CHANGE_DATA_FEED_COLUMN_NAMES.contains(newColumnName)) {
```
[the remark this answers is not in the document]
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1345]: I honestly don't know. As our delta lake connector doesnt support dropping columns I was not able to test dropping a column and then adding new one with the same name.  

On databricks renaming column breaks reading table changes. I got this error:
```

Error in SQL statement: DeltaUnsupportedOperationException: Retrieving table changes between version 0 and 5 failed because of an incompatible schema change.
Your read schema is {"type":"struct","fields":[{"name":"id","type":"integer","nullable":true,"metadata":{"delta.columnMapping.id":1,"delta.columnMapping.physicalName":"id"}},{"name":"new_name","type":"string","nullable":true,"metadata":{"delta.columnMapping.id":2,"delta.columnMapping.physicalName":"name"}}]} at version 5, but we found an incompatible schema change at version 3.
If possible, please query table changes separately from version 0 to 3 - 1, and from version 3 to 5.
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java [hunk 5]
```diff
@@ -1188,6 +1188,158 @@ public Object[][] columnMappingWithTrueAndFalseDataProvider()
         return cartesianProduct(supportedColumnMappingForDmlDataProvider(), trueFalse());
     }
 
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_73, DELTA_LAKE_EXCLUDE_91, DELTA_LAKE_EXCLUDE_104, PROFILE_SPECIFIC_TESTS}, dataProvider = "columnMappingDataProvider")
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:1191]: #18097

**Comment by findepi** [comment]
the build isn't green. is it of a concern?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java [hunk 6]
```diff
@@ -334,6 +333,33 @@ public void testDropNonEmptySchemaWithTable()
...
+                .hasMessageContaining("Cannot rename column with the column mapping: NONE");
+    }
+
+    @Override
+    public void testRenameColumnName(String columnName)
+    {
+        // Override because the connector doesn't support renaming columns with 'none' column mapping
+        // There are some tests in in io.trino.tests.product.deltalake.TestDeltaLakeColumnMappingMode
+        assertThatThrownBy(() -> super.testRenameColumnName(columnName))
+                .hasMessageContaining("Cannot rename column with the column mapping: NONE");
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java]: This is OK for now, but this clearly shows that from user perspective renames are not quite supported, because they need to opt-in for them during table creation.
We should change default column mapping mode for the CREATE statements. (as soon as we're ready with mapping modes support; not within this PR)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 7]
```diff
@@ -1320,7 +1320,7 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
         DeltaLakeTableHandle handle = checkValidTableHandle(tableHandle);
         checkSupportedWriterVersion(session, handle);
         ColumnMappingMode columnMappingMode = getColumnMappingMode(handle.getMetadataEntry());
-        if (changeDataFeedEnabled(handle.getMetadataEntry()) && CHANGE_DATA_FEED_COLUMN_NAMES.contains(newColumnMetadata.getName())) {
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1322]: - this is part of `fixup! Add support for renaming column in Delta Lake` but this changes ADD COLUMN flow; move to separate commit
- update message accordingly. "Column name %s is forbidden when change data feed is enabled" is no longer applicable, since no other name is allowed

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 8]
```diff
@@ -516,6 +516,21 @@ private static <T> Map<String, T> getColumnProperty(String json, Function<JsonNo
         }
     }
 
+    /**
+     * @return the case-sensitive column names
+     */
+    public static List<String> getColumnNames(MetadataEntry metadataEntry)
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:522]: Can this be indicated in the method name?

eg `getExactColumnNames`

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java [hunk 9]
```diff
@@ -336,28 +336,29 @@ public void testDropNonEmptySchemaWithTable()
     @Override
     public void testRenameColumn()
     {
-        // Override because the connector doesn't support renaming columns with 'none' column mapping
-        // There are some tests in in io.trino.tests.product.deltalake.TestDeltaLakeColumnMappingMode
-        assertThatThrownBy(super::testRenameColumn)
-                .hasMessageContaining("Cannot rename column with the column mapping: NONE");
+        // Override because the default column mapping mode 'none' doesn't support renaming columns
+        super.testRenameColumn("WITH (column_mapping_mode = 'id')");
+        super.testRenameColumn("WITH (column_mapping_mode = 'name')");
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java:341]: I preferred the previous version: BCT covering the default setup and separate test methods covering the non-default configuration

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 10]
```diff
@@ -4892,20 +4902,20 @@ public void testRenameColumnName(String columnName)
...
+    protected void testRenameColumnName(String tableProperty, String columnName, boolean delimited)
     {
         String nameInSql = toColumnNameInSql(columnName, delimited);
         String tableName = "tcn_" + nameInSql.replaceAll("[^a-z0-9]", "") + randomNameSuffix();
         // Use complex identifier to test a source column name when renaming columns
         String sourceColumnName = "a;b$c";
 
         try {
-            assertUpdate("CREATE TABLE " + tableName + "(\"" + sourceColumnName + "\" varchar(50))");
+            assertUpdate("CREATE TABLE " + tableName + "(\"" + sourceColumnName + "\" varchar(50))" + tableProperty);
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:4918]: `Optional<String> tableProperties`
if non-empty, append `"WITH (%s)".formatted(tableProperties.get())"`

However, i feel like it would be better not to have this in BCT at all.
Let BCT keep testing default case and let dedicated test methods cover the non-default setup

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 11]
```diff
@@ -1393,6 +1394,84 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
+        MetadataEntry metadataEntry = table.getMetadataEntry();
+        ColumnMappingMode columnMappingMode = getColumnMappingMode(metadataEntry);
+        if (columnMappingMode != ColumnMappingMode.NAME && columnMappingMode != ColumnMappingMode.ID) {
+            throw new TrinoException(NOT_SUPPORTED, "Cannot rename column with the column mapping: " + columnMappingMode);
+        }
+
+        ConnectorTableMetadata tableMetadata = getTableMetadata(session, table);
+        long commitVersion = table.getReadVersion() + 1;
+        List<String> partitionColumns = getPartitionedBy(tableMetadata.getProperties()).stream()
+                .map(columnName -> columnName.equals(sourceColumnName) ? newColumnName : columnName)
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java]: It's covered by `TestDeltaLakeColumnMappingMode.testRenameColumnWithColumnMappingMode`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 12]
```diff
@@ -1393,6 +1394,84 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
+                                        ? Map.entry(newColumnName, DeltaLakeColumnStatistics.create(stats.getValue().getTotalSizeInBytes(), stats.getValue().getNdvSummary()))
+                                        : stats)
+                                .collect(toImmutableMap(Entry::getKey, Entry::getValue)),
+                        existingStatistics.getAnalyzedColumns()
+                                .map(analyzedColumns -> analyzedColumns.stream().map(column -> column.equals(sourceColumnName) ? newColumnName : column).collect(toImmutableSet())));
+                statisticsAccess.updateExtendedStatistics(session, table.getSchemaTableName(), table.getLocation(), statistics);
+            });
+        }
+        catch (Exception e) {
+            throw new TrinoException(DELTA_LAKE_BAD_WRITE, format("Unable to rename '%s' column for: %s.%s", sourceColumnName, table.getSchemaName(), table.getTableName()), e);
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1473]: Now I think we shouldn't update extended statistics because it uses physical column names internally and renaming columns doesn't change physical names. I will remove this block.

**Review by findinpath — APPROVED** [review]

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 13]
```diff
@@ -1393,6 +1394,84 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
+    @Override
+    public void renameColumn(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle, String newColumnName)
+    {
+        DeltaLakeTableHandle table = (DeltaLakeTableHandle) tableHandle;
+        DeltaLakeColumnHandle deltaLakeColumn = (DeltaLakeColumnHandle) columnHandle;
+        String sourceColumnName = deltaLakeColumn.getBaseColumnName();
+
+        checkSupportedWriterVersion(session, table);
+        if (changeDataFeedEnabled(table.getMetadataEntry()) && CHANGE_DATA_FEED_COLUMN_NAMES.contains(newColumnName)) {
+            throw new TrinoException(NOT_SUPPORTED, "Unable to rename to %s columns when change data feed is enabled: %s".formatted(CHANGE_DATA_FEED_COLUMN_NAMES, newColumnName));
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java]: It doesn't depend on mapping mode. There's `spark.databricks.delta.changeDataFeed.unsafeBatchReadOnIncompatibleSchemaChanges.enabled` config property, but I don't think we want to allow unsafe operations.
